Tests/MongoSwiftSyncTests/SyncAuthTests.swift

```diff
@@ -0,0 +1,139 @@
+import Foundation
+import MongoSwift
+import Nimble
+import TestsCommon
+
+/// Represents a test user to create and test authenticating with.
+struct TestUser {
+    let username: String
+    let password: String
+    let mechanisms: [AuthMechanism]
+
+    /// A command to create this user.
+    var createCmd: Document {
+        return [
+            "createUser": .string(self.username),
+            "pwd": .string(self.password),
+            "roles": ["root"],
+            "mechanisms": .array(self.mechanisms.map { .string($0.rawValue) })
+        ]
+    }
+
+    /// Adds this user's username and password, and an optionally provided auth mechanism, to the connection string.
+    func addToConnString(_ connStr: String, mechanism: AuthMechanism? = nil) throws -> String {
+        // find where the first / is.
+        guard let firstSlash = connStr.firstIndex(of: "/") else {
+            throw TestError(message: "expected connection string to contain slash")
+        }
+
+        // this should also be a / in a properly formatted URI.
+        let nextIdx = connStr.index(after: firstSlash)
+        guard connStr[nextIdx] == "/" else {
+            throw TestError(message: "expected connection string to contain '//'")
+        }
+
+        // we want to split right after the // to insert the username and password.
+        let splitIdx = connStr.index(firstSlash, offsetBy: 2)
+
+        let joined = "\(connStr[..<splitIdx])\(self.username):\(self.password)@\(connStr[splitIdx...])"
+        guard let mech = mechanism else {
+            return joined
+        }
+
+        // assume there are already URL parameters if there's a ?, e.g. mongodb://...../?replset=replset0
+        if connStr.contains("?") {
+            return "\(joined)&authMechanism=\(mech.rawValue)"
+        }
+        // assume it is a URI that ends with a / and has no params, e.g. mongodb://localhost:27017/
+        else if connStr.hasSuffix("/") {
+            return "\(joined)?authMechanism=\(mech.rawValue)"
+        }
+        // assume the URI does not end with a / and also has no params, e.g. mongodb://localhost:27017
+        return "\(joined)/?authMechanism=\(mech.rawValue)"
+    }
+}
+
+final class SyncAuthTests: MongoSwiftTestCase {
+    // TODO: SWIFT-640: spec says "Drivers that allow specifying auth parameters in code as well as via connection
+    // string should test both for the test cases described below". Once we support setting auth options via options
+    // struct we should test that here too.
+    func testAuthProseTests() throws {
```

Comment:
the prose tests only use the public API, so I could move them over. 

the JSON tests use the `ConnectionString` type so they need to be in the async test module.